File: app/src/main/java/com/google/samples/apps/topeka/widget/quiz/AbsQuizView.java
/*
 * Copyright 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.samples.apps.topeka.widget.quiz;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.support.annotation.DimenRes;
import android.util.IntProperty;
import android.util.Property;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import android.view.animation.Interpolator;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.samples.apps.topeka.R;
import com.google.samples.apps.topeka.activity.QuizActivity;
import com.google.samples.apps.topeka.model.Category;
import com.google.samples.apps.topeka.model.quiz.Quiz;
import com.google.samples.apps.topeka.widget.fab.CheckableFab;

/**
 * This is the base class for displaying a {@link com.google.samples.apps.topeka.model.quiz.Quiz}.
 * <p>
 * Subclasses need to implement {@link AbsQuizView#createQuizContentView()}
 * in order to allow solution of a quiz.
 * </p>
 * <p>
 * Also {@link AbsQuizView#allowAnswer(boolean)} needs to be called with
 * <code>true</code> in order to mark the quiz solved.
 * </p>
 *
 * @param <Q> The type of {@link com.google.samples.apps.topeka.model.quiz.Quiz} you want to
 * display.
 */
public abstract class AbsQuizView<Q extends Quiz> extends FrameLayout {

    /** Property for animating the foreground color */
    public static final Property<FrameLayout, Integer> FOREGROUND_COLOR =
            new IntProperty<FrameLayout>("foregroundColor") {

                @Override
                public void setValue(FrameLayout object, int value) {
                    if (object.getForeground() instanceof ColorDrawable) {
                        ((ColorDrawable) object.getForeground()).setColor(value);
                    } else {
                        object.setForeground(new ColorDrawable(value));
                    }
                }

                @Override
                public Integer get(FrameLayout object) {
                    return ((ColorDrawable) object.getForeground()).getColor();
                }
            };

    protected final int mMinHeightTouchTarget;
    private final int mSpacingDouble;
    private final LayoutInflater mLayoutInflater;
    private final Category mCategory;
    private final Q mQuiz;
    private final Interpolator mFastOutSlowInInterpolator;
    private final Interpolator mLinearOutSlowInInterpolator;
    private final int mColorAnimationDuration;
    private final int mIconAnimationDuration;
    private final int mScaleAnimationDuration;
    private boolean mAnswered;
    private TextView mQuestionView;
    private CheckableFab mSubmitAnswer;

    /**
     * Enables creation of views for quizzes.
     *
     * @param context The context for this view.
     * @param category The {@link Category} this view is running in.
     * @param quiz The actual {@link Quiz} that is going to be displayed.
     */
    public AbsQuizView(Context context, Category category, Q quiz) {
        super(context);
        mQuiz = quiz;
        mCategory = category;
        mSpacingDouble = getResources().getDimensionPixelSize(R.dimen.spacing_double);
        mSubmitAnswer = getSubmitButton(context);
        mLayoutInflater = LayoutInflater.from(context);
        mMinHeightTouchTarget = getResources()
                .getDimensionPixelSize(R.dimen.min_height_touch_target);
        mFastOutSlowInInterpolator = AnimationUtils
                .loadInterpolator(getContext(), android.R.interpolator.fast_out_slow_in);
        mLinearOutSlowInInterpolator = AnimationUtils
                .loadInterpolator(getContext(), android.R.interpolator.linear_out_slow_in);
        mColorAnimationDuration = 400;
        mIconAnimationDuration = 300;
        mScaleAnimationDuration = 200;

        setId(quiz.getId());
        setUpQuestionView();
        LinearLayout container = createContainerLayout(context);
        View quizContentView = getInitializedContentView();
        addContentView(container, quizContentView);
        addOnLayoutChangeListener(new OnLayoutChangeListener() {
            @Override
            public void onLayoutChange(View v, int left, int top, int right, int bottom,
                    int oldLeft,
                    int oldTop, int oldRight, int oldBottom) {
                removeOnLayoutChangeListener(this);
                addFloatingActionButton();
            }
        });
    }

    /**
     * Sets the behaviour for all question views.
     */
    private void setUpQuestionView() {
        mQuestionView = (TextView) mLayoutInflater.inflate(R.layout.question, this, false);
        mQuestionView.setText(getQuiz().getQuestion());
    }

    private LinearLayout createContainerLayout(Context context) {
        LinearLayout container = new LinearLayout(context);
        container.setId(R.id.absQuizViewContainer);
        container.setOrientation(LinearLayout.VERTICAL);
        return container;
    }

    private View getInitializedContentView() {
        View quizContentView = createQuizContentView();
        quizContentView.setId(R.id.quiz_content);
        quizContentView.setSaveEnabled(true);
        setDefaultPadding(quizContentView);
        if (quizContentView instanceof ViewGroup) {
            ((ViewGroup) quizContentView).setClipToPadding(false);
        }
        setMinHeightInternal(quizContentView, R.dimen.min_height_question);
        return quizContentView;
    }

    private void addContentView(LinearLayout container, View quizContentView) {
        LayoutParams layoutParams = new LayoutParams(LayoutParams.MATCH_PARENT,
                LayoutParams.WRAP_CONTENT);
        container.addView(mQuestionView, layoutParams);
        container.addView(quizContentView, layoutParams);
        addView(container, layoutParams);
    }

    private void addFloatingActionButton() {
        final int fabSize = getResources().getDimensionPixelSize(R.dimen.size_fab);
        int bottomOfQuestionView = findViewById(R.id.question_view).getBottom();
        final LayoutParams fabLayoutParams = new LayoutParams(fabSize, fabSize,
                Gravity.END | Gravity.TOP);
        final int halfAFab = fabSize / 2;
        fabLayoutParams.setMargins(0, // left
                bottomOfQuestionView - halfAFab, //top
                0, // right
                mSpacingDouble); // bottom
        fabLayoutParams.setMarginEnd(mSpacingDouble);
        addView(mSubmitAnswer, fabLayoutParams);
    }

    private CheckableFab getSubmitButton(Context context) {
        if (null == mSubmitAnswer) {
            mSubmitAnswer = new CheckableFab(context);
            mSubmitAnswer.setId(R.id.submitAnswer);
            mSubmitAnswer.setVisibility(GONE);
            mSubmitAnswer.setScaleY(0);
            mSubmitAnswer.setScaleX(0);
            mSubmitAnswer.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    submitAnswer(v);
                }
            });
        }
        return mSubmitAnswer;
    }

    private void setDefaultPadding(View view) {
        view.setPadding(mSpacingDouble, mSpacingDouble, mSpacingDouble, mSpacingDouble);
    }

    protected LayoutInflater getLayoutInflater() {
        return mLayoutInflater;
    }

    /**
     * Implementations should create the content view for the type of
     * {@link com.google.samples.apps.topeka.model.quiz.Quiz} they want to display.
     *
     * @return the created view to solve the quiz.
     */
    protected abstract View createQuizContentView();

    /**
     * Implementations must make sure that the answer provided is evaluated and correctly rated.
     *
     * @return <code>true</code> if the question has been correctly answered, else
     * <code>false</code>.
     */
    protected abstract boolean isAnswerCorrect();

    /**
     * Save the user input to a bundle for orientation changes.
     *
     * @return The bundle containing the user's input.
     */
    public abstract Bundle getUserInput();

    /**
     * Restore the user's input.
     *
     * @param savedInput The input that the user made in a prior instance of this view.
     */
    public abstract void setUserInput(Bundle savedInput);

    public Q getQuiz() {
        return mQuiz;
    }

    protected boolean isAnswered() {
        return mAnswered;
    }

    /**
     * Sets the quiz to answered or unanswered.
     *
     * @param answered <code>true</code> if an answer was selected, else <code>false</code>.
     */
    protected void allowAnswer(final boolean answered) {
        if (null != mSubmitAnswer) {
            final float targetScale = answered ? 1f : 0f;
            if (answered) {
                mSubmitAnswer.setVisibility(View.VISIBLE);
            }
            mSubmitAnswer.animate().scaleX(targetScale).scaleY(targetScale)
                    .setInterpolator(mFastOutSlowInInterpolator);
            mAnswered = answered;
        }
    }

    /**
     * Sets the quiz to answered if it not already has been answered.
     * Otherwise does nothing.
     */
    protected void allowAnswer() {
        if (!isAnswered()) {
            allowAnswer(true);
        }
    }

    /**
     * Allows children to submit an answer via code.
     */
    protected void submitAnswer() {
        submitAnswer(findViewById(R.id.submitAnswer));
    }


    @SuppressWarnings("UnusedParameters")
    private void submitAnswer(final View v) {
        final boolean answerCorrect = isAnswerCorrect();
        mQuiz.setSolved(true);
        performScoreAnimation(answerCorrect);
    }

    /**
     * Animates the view nicely when the answer has been submitted.
     *
     * @param answerCorrect <code>true</code> if the answer was correct, else <code>false</code>.
     */
    private void performScoreAnimation(final boolean answerCorrect) {

        mSubmitAnswer.setChecked(answerCorrect);

        // Decide which background color to use.
        final int backgroundColor = getResources()
                .getColor(answerCorrect ? R.color.green : R.color.red);
        animateFabBackgroundColor(backgroundColor);
        hideFab();
        resizeView();
        moveViewOffScreen(answerCorrect);
        // Animate the foreground color to match the background color.
        // This overlays all content within the current view.
        animateForegroundColor(backgroundColor);
    }

    private void hideFab() {
        mSubmitAnswer.animate()
                .setDuration(mScaleAnimationDuration)
                .setStartDelay(mIconAnimationDuration * 2)
                .scaleX(0f)
                .scaleY(0f)
                .setInterpolator(mLinearOutSlowInInterpolator)
                .start();
    }

    private void resizeView() {
        final float widthHeightRatio = (float) getHeight() / (float) getWidth();

        // Animate X and Y scaling separately to allow different start delays.
        animate()
                .scaleY(.5f / widthHeightRatio)
                .setDuration(300)
                .setStartDelay(750)
                .start();
        animate()
                .scaleX(.5f)
                .setDuration(300)
                .setStartDelay(800)
                .start();
    }

    private void animateFabBackgroundColor(int backgroundColor) {
        // Set color, duration and interpolator for the color change. Then start the animation.
        final ObjectAnimator fabColorAnimator = ObjectAnimator
                .ofArgb(mSubmitAnswer, "backgroundColor", Color.WHITE, backgroundColor);
        fabColorAnimator.setDuration(mColorAnimationDuration)
                .setInterpolator(mFastOutSlowInInterpolator);
        fabColorAnimator.start();
    }

    private void animateForegroundColor(int targetColor) {
        final ObjectAnimator foregroundAnimator = ObjectAnimator
                .ofArgb(this, FOREGROUND_COLOR, Color.WHITE, targetColor);
        foregroundAnimator
                .setDuration(200)
                .setInterpolator(mLinearOutSlowInInterpolator);
        foregroundAnimator.setStartDelay(750);
        foregroundAnimator.start();
    }

    private void moveViewOffScreen(final boolean answerCorrect) {
        // Animate the current view off the screen.
        animate()
                .setDuration(200)
                .setStartDelay(1200)
                .setInterpolator(mLinearOutSlowInInterpolator)
                .withEndAction(new Runnable() {
                    @Override
                    public void run() {
                        mCategory.setScore(getQuiz(), answerCorrect);
                        if (getContext() instanceof QuizActivity) {
                            ((QuizActivity) getContext()).proceed();
                        }
                    }
                })
                .start();
    }

    private void setMinHeightInternal(View view, @DimenRes int resId) {
        view.setMinimumHeight(getResources().getDimensionPixelSize(resId));
    }
}

File: quiz/src/main/java/com/google/samples/apps/topeka/widget/quiz/AbsQuizView.kt
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.samples.apps.topeka.widget.quiz

import android.animation.ArgbEvaluator
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.support.annotation.ColorInt
import android.support.annotation.LayoutRes
import android.support.v4.content.ContextCompat
import android.support.v4.view.MarginLayoutParamsCompat
import android.support.v4.view.animation.LinearOutSlowInInterpolator
import android.util.Property
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import com.google.samples.apps.topeka.activity.QuizActivity
import com.google.samples.apps.topeka.helper.ApiLevelHelper
import com.google.samples.apps.topeka.helper.FOREGROUND_COLOR
import com.google.samples.apps.topeka.helper.onLayoutChange
import com.google.samples.apps.topeka.model.Category
import com.google.samples.apps.topeka.model.quiz.Quiz
import com.google.samples.apps.topeka.quiz.R
import com.google.samples.apps.topeka.widget.fab.CheckableFab

const val KEY_ANSWER = "answer"

/**
 * This is the base class for displaying a [com.google.samples.apps.topeka.model.quiz.Quiz].
 *
 *
 * Subclasses need to implement [AbsQuizView.createQuizContentView]
 * in order to allow solution of a quiz.
 *
 *
 *
 * Also [AbsQuizView.allowAnswer] needs to be called with
 * `true` in order to mark the quiz solved.
 *

 * @param <Q> The type of [com.google.samples.apps.topeka.model.quiz.Quiz] you want to
 * display.
</Q> */
abstract class AbsQuizView<out Q : Quiz<*>>
/**
 * Enables creation of views for quizzes.

 * @param context The context for this view.
 *
 * @param category The [Category] this view is running in.
 *
 * @param quiz The actual [Quiz] that is going to be displayed.
 */
(context: Context, private val category: Category, val quiz: Q) : FrameLayout(context) {

    private val layoutInflater: LayoutInflater = LayoutInflater.from(context)

    val quizContentView: View?

    init {
        id = quiz.id
        val container = createContainerLayout(context)

        quizContentView = createQuizContentView()?.apply {
            id = com.google.samples.apps.topeka.base.R.id.quiz_content
            isSaveEnabled = true
            if (this is ViewGroup) {
                clipToPadding = false
            }
            setMinHeightInternal(this)
            addContentView(container, this)
        }
        onLayoutChange { addFloatingActionButton() }
    }

    private val ANSWER_HIDE_DELAY = 500L

    private val FOREGROUND_COLOR_CHANGE_DELAY = 750L
    private val doubleSpacing = resources.getDimensionPixelSize(
            com.google.samples.apps.topeka.base.R.dimen.spacing_double)
    private val linearOutSlowInInterpolator = LinearOutSlowInInterpolator()
    private val quizHandler = Handler()

    private val submitAnswer = initCheckableFab()

    private fun initCheckableFab(): CheckableFab {
        return (inflate<CheckableFab>(R.layout.answer_submit)).apply {
            hide()
            setOnClickListener {
                submitAnswer()
                if (inputMethodManager.isAcceptingText) {
                    inputMethodManager.hideSoftInputFromWindow(it.windowToken, 0)
                }
                isEnabled = false
            }
        }
    }

    private val inputMethodManager: InputMethodManager =
            context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager

    protected var isAnswered = false
        private set

    /**
     * Implementations must make sure that the answer provided is evaluated and correctly rated.

     * @return `true` if the question has been correctly answered, else
     * `false`.
     */
    protected abstract val isAnswerCorrect: Boolean

    /**
     * Holds the user input for orientation change purposes.
     */
    abstract var userInput: Bundle

    private fun createContainerLayout(context: Context): LinearLayout {
        return LinearLayout(context).apply {
            id = com.google.samples.apps.topeka.base.R.id.absQuizViewContainer
            orientation = LinearLayout.VERTICAL
        }
    }

    private fun addContentView(container: LinearLayout, quizContentView: View) {
        val layoutParams = FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.WRAP_CONTENT)
        container.run {
            addView((layoutInflater.inflate(R.layout.question, this, false) as TextView).apply {
                setBackgroundColor(ContextCompat.getColor(context,
                        category.theme.primaryColor))
                text = quiz.question
            }, layoutParams)
            addView(quizContentView, layoutParams)
        }
        addView(container, layoutParams)
    }

    private fun addFloatingActionButton() {
        val fabSize = resources.getDimensionPixelSize(
                com.google.samples.apps.topeka.base.R.dimen.size_fab)
        val bottomOfQuestionView = findViewById<View>(R.id.question_view).bottom
        val fabLayoutParams = FrameLayout.LayoutParams(fabSize, fabSize, Gravity.END or Gravity.TOP)
        val halfAFab = fabSize / 2
        fabLayoutParams.setMargins(0, // left
                bottomOfQuestionView - halfAFab, // top
                0, // right
                doubleSpacing) // bottom
        MarginLayoutParamsCompat.setMarginEnd(fabLayoutParams, doubleSpacing)
        if (ApiLevelHelper.isLowerThan(Build.VERSION_CODES.LOLLIPOP)) {
            // Account for the fab's emulated shadow.
            fabLayoutParams.topMargin -= submitAnswer.paddingTop / 2
        }
        addView(submitAnswer, fabLayoutParams)
    }

    /**
     * Implementations should create the content view for the type of
     * [com.google.samples.apps.topeka.model.quiz.Quiz] they want to display.

     * @return the created view to solve the quiz.
     */
    protected abstract fun createQuizContentView(): View?

    /**
     * Sets the quiz to answered or unanswered.

     * @param answered `true` if an answer was selected, else `false`.
     */
    protected fun allowAnswer(answered: Boolean = true) {
        with(submitAnswer) { if (answered) show() else hide() }
        isAnswered = answered
    }

    /**
     * Allows children to submit an answer via code.
     */
    protected open fun submitAnswer() {
        quiz.solved = true
        performScoreAnimation(isAnswerCorrect)
    }

    /**
     * Animates the view when the answer has been submitted.

     * @param answerCorrect `true` if the answer was correct, else `false`.
     */
    private fun performScoreAnimation(answerCorrect: Boolean) {
        (context as QuizActivity).lockIdlingResource()
        // Decide which background color to use.
        val backgroundColor = ContextCompat.getColor(context,
                if (answerCorrect) com.google.samples.apps.topeka.base.R.color.green
                else com.google.samples.apps.topeka.base.R.color.red)
        adjustFab(answerCorrect, backgroundColor)
        resizeView()
        moveViewOffScreen(answerCorrect)
        // Animate the foreground color to match the background color.
        // This overlays all content within the current view.
        animateForegroundColor(backgroundColor)
    }

    @SuppressLint("NewApi")
    private fun adjustFab(answerCorrect: Boolean, backgroundColor: Int) {
        with(submitAnswer) {
            isChecked = answerCorrect
            backgroundTintList = ColorStateList.valueOf(backgroundColor)
        }
        quizHandler.postDelayed({ submitAnswer.hide() }, ANSWER_HIDE_DELAY)
    }

    private fun resizeView() {
        val widthHeightRatio = height.toFloat() / width.toFloat()
        // Animate X and Y scaling separately to allow different start delays.
        // object animators for x and y with different durations and then run them independently
        resizeViewProperty(View.SCALE_X, .5f, 200)
        resizeViewProperty(View.SCALE_Y, .5f / widthHeightRatio, 300)
    }

    private fun resizeViewProperty(property: Property<View, Float>,
                                   targetScale: Float, durationOffset: Int) {
        with(ObjectAnimator.ofFloat(this, property, 1f, targetScale)) {
            interpolator = linearOutSlowInInterpolator
            startDelay = (FOREGROUND_COLOR_CHANGE_DELAY + durationOffset)
            start()
        }
    }

    override fun onDetachedFromWindow() {
        quizHandler.removeCallbacksAndMessages(null)
        super.onDetachedFromWindow()
    }

    private fun animateForegroundColor(@ColorInt targetColor: Int) {
        with(ObjectAnimator.ofInt(this, FOREGROUND_COLOR, Color.TRANSPARENT, targetColor)) {
            setEvaluator(ArgbEvaluator())
            startDelay = FOREGROUND_COLOR_CHANGE_DELAY
            start()
        }
    }

    private fun moveViewOffScreen(answerCorrect: Boolean) {
        category.setScore(quiz, answerCorrect)
        // Move the current view off the screen.
        quizHandler.postDelayed(this::proceed, FOREGROUND_COLOR_CHANGE_DELAY * 2)
    }

    private fun setMinHeightInternal(view: View) {
        view.minimumHeight = resources.getDimensionPixelSize(
                com.google.samples.apps.topeka.base.R.dimen.min_height_question)
    }

    private fun proceed() {
        if (context is QuizActivity) (context as QuizActivity).proceed()
    }

    protected fun <T: View> inflate(@LayoutRes resId: Int) =
            layoutInflater.inflate(resId, this, false) as T

}